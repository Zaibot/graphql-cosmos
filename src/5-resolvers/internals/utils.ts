import { FeedResponse } from '@azure/cosmos'
import { GraphQLCosmosPageOutput } from '../../4-resolver-builder/3-typedefs-transformer'
import { Lazy } from '../../typescript'
import { SourceDescriptor } from '../x-descriptors'

export const emptyPageResponse: GraphQLCosmosPageOutput<any> = {
  cursor: null,
  nextCursor: null,
  page: [],
  total: 0,
}

export function graphqlCosmosPageResponse<T extends { id: string }>(
  cursor: string | null,
  query: Lazy<Promise<FeedResponse<T>>>,
  count: Lazy<Promise<FeedResponse<number>>>,
  transformer: (item: T) => SourceDescriptor.Embedded<T>
): GraphQLCosmosPageOutput<T> {
  return {
    cursor: cursor,
    nextCursor: () => query().then((x) => x.continuationToken),
    page: () => query().then((x) => (x.resources ?? []).map(transformer)),
    total: () => count().then((x) => x.resources[0]),
  }
}
