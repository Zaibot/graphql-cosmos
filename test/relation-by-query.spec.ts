import gql from 'graphql-tag'
import { printSchemaWithDirectives } from 'graphql-tools'
import { createUnitTestContext } from './utils'

describe(`Reference to other container`, () => {
  const dummyTypeDefs = gql`
    type Query {
      dummies: [Dummy!]! @cosmos(database: "Test", container: "Dummies")
    }

    type Dummy {
      id: ID!
      related: [Related!]! @cosmos(database: "Test", container: "Relations", ours: "relatedIds", pagination: "on")
    }

    type Related {
      id: ID!
      dummies: [Dummy!]! @cosmos(database: "Test", container: "Dummies", theirs: "relatedIds")
    }
  `

  const responses = {
    Dummies: {
      'SELECT VALUE COUNT(1) FROM c': [3],
      'SELECT c.id FROM c ORDER BY c.id': [{ id: `1` }, { id: `2` }, { id: `3` }],
      'SELECT c.id, c.relatedIds FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=1,2,3)': [
        { id: `1`, relatedIds: [`1b`] },
        { id: `2`, relatedIds: [`2b`] },
        { id: `3`, relatedIds: [`3b`] },
      ],

      'SELECT VALUE COUNT(1) FROM c WHERE ARRAY_CONTAINS(c.relatedIds, @p2) (@p2=1b)': [1],
      'SELECT VALUE COUNT(1) FROM c WHERE ARRAY_CONTAINS(c.relatedIds, @p2) (@p2=2b)': [1],
      'SELECT VALUE COUNT(1) FROM c WHERE ARRAY_CONTAINS(c.relatedIds, @p2) (@p2=3b)': [1],

      'SELECT c.id FROM c WHERE ARRAY_CONTAINS(c.relatedIds, @p2) ORDER BY c.id (@p2=1b)': [{ id: `1` }],
      'SELECT c.id FROM c WHERE ARRAY_CONTAINS(c.relatedIds, @p2) ORDER BY c.id (@p2=2b)': [{ id: `2` }],
      'SELECT c.id FROM c WHERE ARRAY_CONTAINS(c.relatedIds, @p2) ORDER BY c.id (@p2=3b)': [{ id: `3` }],
    },
    Relations: {
      'SELECT c.id, c.dummies FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=1b,2b,3b)': [
        { id: `1b` },
        { id: `2b` },
        { id: `3b` },
      ],

      'SELECT VALUE COUNT(1) FROM c WHERE ARRAY_CONTAINS(@p2, c.id) (@p2=1b)': [1],
      'SELECT VALUE COUNT(1) FROM c WHERE ARRAY_CONTAINS(@p2, c.id) (@p2=2b)': [1],
      'SELECT VALUE COUNT(1) FROM c WHERE ARRAY_CONTAINS(@p2, c.id) (@p2=3b)': [1],

      'SELECT c.id FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=1b)': [{ id: `1b` }],
      'SELECT c.id FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=2b)': [{ id: `2b` }],
      'SELECT c.id FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=3b)': [{ id: `3b` }],
    },
  }

  const uc = createUnitTestContext(dummyTypeDefs, responses)

  it(`expects schema to remain the same`, () => {
    const output = printSchemaWithDirectives(uc.schema)
    expect(output).toMatchSnapshot()
  })

  it(`expects meta schema to remain the same`, () => {
    const output = uc.metaSchema
    expect(output).toMatchSnapshot()
  })

  it(`should be retrieve all items (root)`, async () => {
    const result = await uc.execute(`query { dummies { total page { __typename id } } } `)
    if (result.errors?.length) {
      result.errors.forEach((e) => console.error(e.path.join(`/`), e.stack))
    }

    expect(result).toEqual({
      data: {
        dummies: {
          total: 3,
          page: [
            { __typename: 'Dummy', id: `1` },
            { __typename: 'Dummy', id: `2` },
            { __typename: 'Dummy', id: `3` },
          ],
        },
      },
    })
  })

  it(`should be retrieve all items (ours)`, async () => {
    const result = await uc.execute(
      `query { dummies { total page { __typename id related { total page { __typename id } } } } } `
    )
    if (result.errors?.length) {
      result.errors.forEach((e) => console.error(e.path.join(`/`), e.stack))
    }

    expect(result).toEqual({
      data: {
        dummies: {
          total: 3,
          page: [
            {
              __typename: 'Dummy',
              id: `1`,
              related: {
                total: 1,
                page: [{ __typename: 'Related', id: `1b` }],
              },
            },
            {
              __typename: 'Dummy',
              id: `2`,
              related: {
                total: 1,
                page: [{ __typename: 'Related', id: `2b` }],
              },
            },
            {
              __typename: 'Dummy',
              id: `3`,
              related: {
                total: 1,
                page: [{ __typename: 'Related', id: `3b` }],
              },
            },
          ],
        },
      },
    })
  })

  it(`should be retrieve all items (theirs)`, async () => {
    const result = await uc.execute(
      `query { dummies { total page { __typename id related { total page { __typename id dummies { total page { __typename id } } } } } } } `
    )
    if (result.errors?.length) {
      result.errors.forEach((e) => console.error(e.path.join(`/`), e.stack))
    }

    expect(result).toEqual({
      data: {
        dummies: {
          total: 3,
          page: [
            {
              __typename: 'Dummy',
              id: `1`,
              related: {
                total: 1,
                page: [
                  {
                    __typename: 'Related',
                    id: `1b`,
                    dummies: {
                      total: 1,
                      page: [{ __typename: 'Dummy', id: `1` }],
                    },
                  },
                ],
              },
            },
            {
              __typename: 'Dummy',
              id: `2`,
              related: {
                total: 1,
                page: [
                  {
                    __typename: 'Related',
                    id: `2b`,
                    dummies: {
                      total: 1,
                      page: [{ __typename: 'Dummy', id: `2` }],
                    },
                  },
                ],
              },
            },
            {
              __typename: 'Dummy',
              id: `3`,
              related: {
                total: 1,
                page: [
                  {
                    __typename: 'Related',
                    id: `3b`,
                    dummies: {
                      total: 1,
                      page: [{ __typename: 'Dummy', id: `3` }],
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    })
  })
})
