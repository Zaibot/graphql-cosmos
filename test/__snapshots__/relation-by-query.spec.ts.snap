// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reference to other container expects meta schema to remain the same 1`] = `
Object {
  "types": Array [
    Object {
      "container": null,
      "database": null,
      "fields": Array [
        Object {
          "container": "Dummies",
          "database": "Test",
          "fieldname": "dummies",
          "kind": "many-root",
          "ours": null,
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Dummy",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "typename": "Query",
    },
    Object {
      "container": null,
      "database": null,
      "fields": Array [
        Object {
          "container": null,
          "database": null,
          "fieldname": "id",
          "kind": "embedded",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": "Relations",
          "database": "Test",
          "fieldname": "related",
          "kind": "many-ours",
          "ours": "relatedIds",
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Related",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "typename": "Dummy",
    },
    Object {
      "container": null,
      "database": null,
      "fields": Array [
        Object {
          "container": null,
          "database": null,
          "fieldname": "id",
          "kind": "embedded",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": "Dummies",
          "database": "Test",
          "fieldname": "dummies",
          "kind": "many-theirs",
          "ours": null,
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Dummy",
          "sortOurs": null,
          "sortable": false,
          "theirs": "relatedIds",
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "typename": "Related",
    },
  ],
}
`;

exports[`Reference to other container expects schema to remain the same 1`] = `
"schema {
  query: Query
}
type Query {
  dummies(cursor: String, limit: Int): DummyPage!
}
type Dummy {
  id: ID!
  related(cursor: String, limit: Int): RelatedPage!
}
type Related {
  id: ID!
  dummies(cursor: String, limit: Int): DummyPage!
}
type DummyPage {
  total: Int!
  cursor: String
  nextCursor: String
  page: [Dummy]!
}
type RelatedPage {
  total: Int!
  cursor: String
  nextCursor: String
  page: [Related]!
}"
`;
